scripts/version_buckets.py

```diff
@@ -0,0 +1,243 @@
+def main(verbose=False):
+    """
+    A tool to manage the versioning and life cycle of S3 buckets.
+    """
+    if verbose:
+        global VERBOSE
+        VERBOSE = True
+        verbose_handler = logging.StreamHandler(sys.stdout)
+        verbose_handler.setLevel(logging.INFO)
+        verbose_handler.setFormatter(log_formatter)
+        logger.addHandler(verbose_handler)
+        click.echo("Logging to {}\n".format(LOGFILE))
+
+
+@main.command()
+@click.option("-n", "--number", required=False, type=int)
+def list(number=0):
+    """
+    List n (or all, if not given) bucket details.
+    """
+    buckets = get_buckets(number)
+    data = [("Name", "Versioned", "LifeCycle")]
+    col_width = [0, 0, 0]
+    rows = []
+    for name, bucket in buckets.items():
+        v = bucket["versioning"]
+        l = bucket["lifecycle"]
+        v = v if v else "Disabled"
+        if l:
+            l = json.dumps(l, indent=1)
+        else:
+            l = "None"
+        data.append((name, v, l))
+    for row in data:
+        for i, info in enumerate(row):
+            col_width[i] = min(max(len(info) + 2, col_width[i]), 48)
+    dashes = tuple(("-" * (width - 1) for width in col_width))
+    data.insert(1, dashes)
+    click.echo(f"The status of the buckets:")
+    for row in data:
+        output = ""
+        for i in range(3):
+            output += row[i].ljust(col_width[i])
+        if not VERBOSE:
+            click.echo(output)
+        logger.info(output)
+
+
+@main.command()
+@click.option("-n", "--number", required=False, type=int)
+@click.option("-x", "--execute", is_flag=True)
+def update(number=0, execute=False):
+    """
+    Update n (or all, if not given) bucket details. Use --execute to make the
+    changes happen, or else a preview will take place.
+    """
+    buckets = get_buckets(number)
+    for k, v in buckets.items():
+        name = k
+        bucket = v["bucket"]
+        versioning = v["versioning"] == "Enabled"
+        lifecycle = v["lifecycle"]
+        update_bucket(name, bucket, versioning, lifecycle, execute)
+
+
+def update_bucket(name, bucket, versioning, lifecycle, execute):
+    """
+    Update the bucket to the expected versioning and life-cycle settings.
+    """
+    msg = f"Working on {name}."
+    logger.info(msg)
+    if not VERBOSE:
+        click.echo(f"Working on {name}.")
+    # Add versioning if not already set.
+    if not versioning:
+        msg = f"Enabling versioning for {name}."
+        logger.info(msg)
+        if not VERBOSE:
+            click.echo(msg)
+        if execute:
+            # TODO: Uncomment to actually make the changes.
+            # v = bucket.Versioning()
+            # v.enable()
+            click.secho("OK", fg="green")
+        else:
+            click.secho("OK", fg="yellow")
+    # Set life cycle rule to send non-current versions of files to glacier
+    # storage after 30 days.
+    lifecycle_id = f"{name}_lifecycle_configuration"
```

Comment:
Does the ID matter? I don't know much about S3 lifecycle configurations but I wonder if having the name of the bucket in it could cause some problems (e.g. name too long or something).

Additionally it may make sense to have a more descriptive name, e.g. `non_current_versions_to_glacier_after_30d` or something. 

Sorry I only thought about this now, to be honest if it works it doesn't matter, I'm just thinking aloud really ☺️

Reply:
It works. The name as it stands describes what it refers to and what it actually is (a lifecycle configuration). I'm calling a spade, a spade. If I were looking at resources in our AWS console and found this, I think I'd have a pretty good idea what it was, to what it pertained and what it represented. If I wanted to know what it did, I could just look at the definition. ;-)

Reply:
No big deal, but I agree that the bucket name is redundant in the lifecycle ID. It's stored against the bucket, so no need to make it globally unique. Better that the ID front-loads what it does.

Reply:
Ack... I'll just make it `lifecycle_configuration`.